src/Escargot.php

```diff
@@ -306,37 +285,51 @@ public function crawl(): void
         $this->processResponses($this->prepareResponses());
     }
 
+    /**
+     * Adds an URI to the queue if not present already.
+     *
+     * @return bool True if it was added and false if it existed already before
+     */
+    public function addUriToQueue(UriInterface $uri, CrawlUri $foundOn, bool $processed = false): bool
+    {
+        $crawlUrl = $this->queue->get($this->jobId, $uri);
+        if (null === $crawlUrl) {
+            $crawlUrl = new CrawlUri($uri, $foundOn->getLevel() + 1, $processed, $foundOn->getUri());
+            $this->queue->add($this->jobId, $crawlUrl);
+
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Logs a message to the logger if one was provided.
+     */
+    public function log(string $level, string $message, array $context = []): void
+    {
+        if (null === $this->logger) {
+            return;
+        }
+
+        $this->logger->log($level, $message, $context);
+    }
+
     private function processResponses(array $responses): void
     {
         foreach ($this->getClient()->stream($responses) as $response => $chunk) {
             try {
-                if ($chunk->isFirst()) {
-                    if (200 !== $response->getStatusCode()) {
-                        --$this->runningRequests;
-                        $response->cancel();
-                        $this->getEventDispatcher()->dispatch(new UnsuccessfulResponseEvent($this, $response));
-                        continue;
-                    }
-
-                    // We're an HTML crawler, so we reject everything that's not text/html immediately
-                    if (!isset($response->getHeaders()['content-type'][0])
-                        || false === strpos($response->getHeaders()['content-type'][0], 'text/html')
-                    ) {
-                        // TODO: another event?
-                        --$this->runningRequests;
-                        $response->cancel();
-                        continue;
-                    }
+                // Dispatch event
+                $this->getEventDispatcher()->dispatch(new ResponseEvent($this, $response, $chunk));
+
+                // Response was canceled by listener
+                if ('Response has been canceled.' === $response->getInfo('error')) {
```

Comment:
This here @nicolas-grekas :)

Reply:
/cc @seldaek we talked about that recently.
@Toflar I'd accept a PR to add a "canceled" info in the contract and in the component if you're up to.

Reply:
Sure! Will submit next week. master or 4.3?